jobs: separate state machine job logic from scheduling

spaskob opened the pull request:
This is a preparation for making OnFailOrCancel fault tolerant.
Added a new function stepThroughStateMachine that implements the
job state machine transition.
Also some other minor improvements:
 - added a CurrentStatus method to get the job status
 - removed StartJob function as it is hard to enforce not racing
   with adopt jobs.
 - new jobs are registered before they are inserted in the jobs
   table to prevent race b/w CreateAndStartJob and maybeAdoptJobs
 - added V2 logging to resume to be able to debug scheduling problems.
                                 
 Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44097)
<!-- Reviewable:end -->

spaskob committed:
jobs: separate state machine job logic from scheduling

This is a preparation for making OnFailOrCancel fault tolerant.
Added a new function stepThroughStateMachine that implements the
job state machine transtion.
Also some other minor improvements:
 - added a CurrentStatus method to get the job status
 - removed StartJob function as it is hard to enforce not racing
   with adopt jobs.
 - new jobs are registered before they are inserted in the jobs
   table to prevent race b/w CreateAndStartJob and maybeAdoptJobs
 - added V2 logging to resume to be able to debug scheduling probems.

Release note: none

thoszhang reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44097)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @spaskob)

___
*[pkg/jobs/jobs.go, line 569 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDYsXB5_6B6L5_tFyR:-LzDYsXC8LnscZ3p3mWc:b38ra05) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L569)):*
> ```Go
> 	if err := j.runInTxn(ctx, func(ctx context.Context, txn *client.Txn) error {
> 		const selectStmt = "SELECT status FROM system.jobs WHERE id = $1"
> 		row, err := j.registry.ex.QueryRow(ctx, "log-job", txn, selectStmt, *j.id)
> ```

nit: should this have a different name?
___
*[pkg/jobs/jobs.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDWeBh78dq69WXP7XR:-LzDWeBh78dq69WXP7XS:bjovtyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L571)):*
> ```Go
> 		row, err := j.registry.ex.QueryRow(ctx, "log-job", txn, selectStmt, *j.id)
> 		if err != nil {
> 			return errors.Wrapf(err, "job %d: can't query system.jobs", j.id)
> ```

This should be `*j.id` as well.
___
*[pkg/jobs/jobs.go, line 574 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDZ4nrDTOCq88v75kG:-LzDZ4nrDTOCq88v75kH:be7ls92) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L574)):*
> ```Go
> 		}
> 		if row == nil {
> 			return errors.Errorf("job %d: no found in system.jobs", *j.id)
> ```

`not found`
___
*[pkg/jobs/jobs.go, line 578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDZEAi9s3c4OknBkTd:-LzDZEAi9s3c4OknBkTe:b-xmeuof) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L578)):*
> ```Go
> 
> 		var ok bool
> 		statusString, ok = row[0].(*tree.DString)
> ```

I think this can use `tree.MustBeDString()`. The error should at least be an assertion error.
___
*[pkg/jobs/jobs_test.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDnC1m9JgLfb5DEtCW:-LzDZZPt7Zw3o7LeFbK5:b8prs7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs_test.go#L144)):*
> ```Go
> 		jobs.DefaultAdoptInterval = oldInterval
> 	}(jobs.DefaultAdoptInterval)
> 	jobs.DefaultAdoptInterval = time.Millisecond
> ```

1 ms seems kind of low, even for a test. Doesn't that mean we have to query system.jobs 1000 times per second?
___
*[pkg/jobs/registry.go, line 628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDk28m6yJToSjFd0hu:-LzDk28m6yJToSjFd0hv:bb7c8te) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L628)):*
> ```Go
> 
> // stepThroughStateMachine implements the state machine of the job lifecycle.
> // The job is executed with the ctx, so ctx must only by canceled if the job
> ```

`be`
___
*[pkg/jobs/registry.go, line 642 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDsR8P04RptrhDYdwM:-LzDsR8P04RptrhDYdwN:bytlzwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L642)):*
> ```Go
> ) error {
> 	switch status {
> 	case StatusRunning:
> ```

Should we log each state transition?
___
*[pkg/jobs/registry.go, line 644 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDkj_u6Wk7ebXN4WCn:-LzDkj_u6Wk7ebXN4WCo:b9xk2e6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L644)):*
> ```Go
> 	case StatusRunning:
> 		if jobErr != nil {
> 			return errors.Wrap(jobErr, "unexpected error provided for a running job")
> ```

These should be assertion errors (with `NewAssertionErrorWithWrappedErrf`), I thi
___
*[pkg/jobs/registry.go, line 669 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDkXRH9MIvJhYtrKDS:-LzDkXRH9MIvJhYtrKDT:b-70fzgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L669)):*
> ```Go
> 	case StatusSucceeded:
> 		if jobErr != nil {
> 			return errors.Wrap(jobErr, "unexpected error provied for a succeddful job")
> ```

`provided` `successful`
___
*[pkg/jobs/registry.go, line 907 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDsLTnAj-ugVy04joC:-LzDsLTnAj-ugVy04joD:b-bogdzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L907)):*
> ```Go
> }
> 
> // register an about to be resumed job in memory so that it can be killed and
> ```

nit: I think the convention is `register registers`...


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44097)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)



<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44097)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/jobs/jobs.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDWeBh78dq69WXP7XR:-LzIFWUjCtx-9P-JiLQk:bj7cyx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L571)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

This should be `*j.id` as well.
</blockquote></details>

good catch!
I change all uses of j.id to j.ID() as well for consistency
___
*[pkg/jobs/jobs.go, line 574 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDZ4nrDTOCq88v75kG:-LzIFfuqDxdTRQGG7OdD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L574)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`not found`
</blockquote></details>

Done.
___
*[pkg/jobs/jobs.go, line 578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDZEAi9s3c4OknBkTd:-LzIGGBz6eMCvDLjsfig:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs.go#L578)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this can use `tree.MustBeDString()`. The error should at least be an assertion error.
</blockquote></details>

Done.
___
*[pkg/jobs/jobs_test.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDnC1m9JgLfb5DEtCW:-LzIGISd6Ta3qTuDwhiP:b-y9a37b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/jobs_test.go#L144)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

1 ms seems kind of low, even for a test. Doesn't that mean we have to query system.jobs 1000 times per second?
</blockquote></details>

in some tests we even use nanosecond, the intention is to be able to catch subtle races between CreateAndStartJob and maybeAdoptJobs
___
*[pkg/jobs/registry.go, line 628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDk28m6yJToSjFd0hu:-LzIHST90-8YU2R5Zd9h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L628)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`be`
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 642 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDsR8P04RptrhDYdwM:-LzIHh7P2VuP4zOztaX5:b-mvpm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L642)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Should we log each state transition?
</blockquote></details>

good idea
done
___
*[pkg/jobs/registry.go, line 644 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDkj_u6Wk7ebXN4WCn:-LzIIslGEAfF4JlFMyhT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L644)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

These should be assertion errors (with `NewAssertionErrorWithWrappedErrf`), I thi
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 669 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44097#-LzDkXRH9MIvJhYtrKDS:-LzIbMn49PjPC8OAPjvI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3dd981de23ba7271545a92e43dd9bfa2c86067/pkg/jobs/registry.go#L669)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`provided` `successful`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spaskob commented:
bors r+
